BTA-1674-Add-Ability-In-Our-SDK-To-Parse-And-Validate-Webhook-Headers-Ruby

linniethepooh opened the pull request:
Added headers and url to the webhook content example and made used of the `validate_webhook_request` method by validating the webhook headers before allowing them to be parsed.

https://github.com/bitpesa/bitpesa-api-specification/pull/99

sztupy commented on Gemfile.lock:
```diff
@@ -9,6 +9,11 @@ GEM
   remote: https://rubygems.org/
   specs:
     ZenTest (4.11.1)
+    activesupport (5.2.3)
```
note that this gem is going to be used in the wild, so the less dependencies the better. This at the moment would for example tie this gem to Rails 5.2.

Also note as far as I know Gemfile.lock should not be commited for public gems, only the Gemfile itself

archmagos replied:
Adding to this, your build will fail unless you remove the lockfile. Your Gemfile specifies ZenTest 4.11.2 which is necessary to pass Travis, but the lockfile sets ZenTest to 4.11.1. It's an issue that  [this OpenApi release](https://github.com/OpenAPITools/openapi-generator/releases/tag/v4.0.0-beta3) will fix, but for now make sure the build is using v4.11.2.

linniethepooh replied:
thanks guys, changes made!

sztupy commented on example/client.rb:
```diff
@@ -212,47 +213,63 @@ def create_and_fund_transaction_example
-          "input_currency": "GBP",
-          "id": "cd390e83-a9de-404d-a660-4da0f1342c04",
-          "metadata": {},
+    headers = <<-JSON
```
no need to have this as a JSON string - we can expect that the user will get the headers in a nice format from the library used.

sztupy replied:
Also we only need the nonce, key and secret here, not the rest

sztupy commented on example/client.rb:
```diff
@@ -397,26 +414,38 @@ def webhook_parse_example
+    webhook_url = webhook_content.url
+    body = webhook_content.body
+
+    if webhook_api.validate_webhook_request(webhook_url, body, nonce, signature)
```
we might need to test as well that they key we got is the same key we store for the client

sztupy replied:
Also where was this `validate_webhook_request` method defined?

emanuelegorga replied:
Hey Zsolt I have been pairing with Linnie on this - the method `validate_webhook_request` is contained in api-specification project. Here is the link of the PR: [API-SPECIFICATION](https://github.com/bitpesa/bitpesa-api-specification/pull/99)

linniethepooh replied:
tested the above, and liked the other PR in the commit message.

sztupy commented on example/client.rb:
```diff
@@ -397,26 +414,38 @@ def webhook_parse_example
       }
     JSON
 
+    webhook_content = ResponseStruct.new(
```
This also looks a bit overcomplicated for this example. Do we need this?

sztupy commented on example/client.rb:
```diff
@@ -397,26 +414,38 @@ def webhook_parse_example
-      puts document_webhook.object.to_s
-    else
-      puts webhook
+    webhook = webhook_api.parse_response(webhook_content.body, "Webhook")
```
we can move this line after the first if. No need to parse the response if it doesn't even validate our security - in fact we shouldn't touch it before validating the headers for security

sztupy commented on example/client.rb:
```diff
@@ -397,26 +414,38 @@ def webhook_parse_example
-    else
-      puts webhook
+    webhook = webhook_api.parse_response(webhook_content.body, "Webhook")
+    nonce = JSON.parse(webhook_content.headers)['Authorization-Nonce']
```
as above: no need for this complex setup. It's enough to have nonce = headers['Authorization-Nonce'], etc. We can expect that the headers will be in a nice format here as most client libraries will return them in a hash, and never in JSON

linniethepooh replied:
change made

archmagos reviewed (commented):
Your build is failing as you still have the lockfile committed, which is causing conflicts. If you remove it from the project and then commit that change alongside your `client.rb` file, the build should pass.

Commit:
BTA-1674-Add-Ability-In-Our-SDK-To-Parse-And-Validate-Webhook-Headers-Ruby

Added headers and url to the webhook content example and made used of the
`validate_webhook_request` method which validates the webhook headers before
allowing them to be parsed.
